### PR TITLE
Reset user password

### PR DESCRIPTION
**Pull request by Blackfry**
duplicate commits here from when I rebased

### CODE REVIEW

File: app/components/ResetPassword/RequestResetEmailForm.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import React from 'react'
+import { Form } from 'react-redux-form'
+import ErrorBox from '../../components/shared/form/ErrorBox'
+import Textfield from '../../components/shared/form/Textfield'
+import PageAlert from '@gov.au/page-alerts'
+import { required, validEmail } from '../validators'
+
+export const RequestResetEmailForm = ({ model, form, user, submitClicked, handleSubmit }) => {
```
Comment (matt-sm) [inline app/components/ResetPassword/RequestResetEmailForm.js]: the pattern I propose we follow:
```
const RequestRestEmailForm = props => {
  const { model, form, user, submitClicked, handleSubmit } = props
  return (
    // JSX
  )
}

RequestResetEmailForm.propTypes = {
  // type for each prop
}

export default RequestResetEmailForm
```

File: app/pages/ResetPasswordPage.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        form={form}
+        user={user}
+        submitClicked={this.onSubmitClicked}
+        handleSubmit={handleSubmit}
+      />
+    )
+  }
+}
+
+RequestResetEmail.propTypes = {
```
Comment (matt-sm) [inline app/pages/ResetPasswordPage.js]: missing props

File: app/pages/ResetPasswordPage.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+import { connect } from 'react-redux'
+import BaseForm from '../components/shared/form/BaseForm'
+import formProps from '../components/shared/form/formPropsSelector'
+import { RequestResetEmailForm } from '../components/ResetPassword/RequestResetEmailForm'
+import { ResetPasswordForm } from '../components/ResetPassword/ResetPasswordForm'
+import { sendResetPasswordEmail, submitResetPassword, getUserDataFromToken } from '../actions/resetPasswordActions'
+import { setErrorMessage } from '../actions/appActions'
+
+class RequestResetEmail extends BaseForm {
+  static propTypes = {
```
Comment (matt-sm) [inline app/pages/ResetPasswordPage.js]: kill these static prototypes - you've got them down below

File: app/pages/ResetPasswordPage.js [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+const mapDispatchToProps = dispatch => {
+  return {
+    handleSubmit: model => dispatch(sendResetPasswordEmail(model))
+  }
+}
+
+export const RequestResetEmailContainer = withRouter(connect(mapStateToProps, mapDispatchToProps)(RequestResetEmail))
+
+class ResetPassword extends BaseForm {
```
Comment (matt-sm) [inline app/pages/ResetPasswordPage.js]: separate file?

File: app/pages/ResetPasswordPage.js [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+const mapResetDispatchToProps = dispatch => {
+  return {
+    handleSubmit: payload => dispatch(submitResetPassword(payload)),
+    loadInitialData: token => dispatch(getUserDataFromToken(token)),
+    passwordsMatchMessage: message => dispatch(setErrorMessage(message))
+  }
+}
+
+export const ResetPasswordContainer = withRouter(connect(mapResetStateToProps, mapResetDispatchToProps)(ResetPassword))
```
Comment (matt-sm) [inline app/pages/ResetPasswordPage.js]: I don't get where this is supposed to fit in - shouldn't it be part of the page?

File: app/routes.js [hunk 6]
```diff
@@ -14,12 +15,13 @@ const RootContainer = () =>
...
 
-export const Routes = ({ match }) =>
+export const Routes = () =>
   <Switch>
-    <Route exact path={match.url} component={SignupContainer} />
+    <Route exact path={rootPath} component={SignupContainer} />
     <Route path={`${rootPath}/signup`} component={SignupContainer} />
     <Route path={`${rootPath}/createuser/:tokenstring`} component={CreateUserPage} />
     <Route path={`${rootPath}/brief/:brief_id/respond`} component={BriefResponse} />
+    <Route path={`${rootPath}/reset-password`} component={ResetPasswordPage} />
```
Comment (matt-sm) [inline app/routes.js]: we've got dashes in this route but not createuser?

File: app/pages/ResetPasswordPage.js [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+          <Route
+            path={`${match.url}/:tokenString`}
+            render={() =>
+              <ResetPasswordForm
+                model={model}
+                form={form}
+                user={user}
+                submitClicked={this.onSubmitClicked}
+                handleSubmit={values => this.handleResetPasswordSubmit(values)}
```
Comment (matt-sm) [inline app/pages/ResetPasswordPage.js]: why is this function on the class where the other `handleSubmit` was a prop?

File: app/pages/ResetPasswordPage.js [hunk 8]
```diff
@@ -0,0 +1,119 @@
+import React, { Component } from 'react'
```
Comment (matt-sm) [inline app/pages/ResetPasswordPage.js]: I like it 👍

File: app/routes.js [hunk 9]
```diff
@@ -3,24 +3,19 @@ import { withRouter, Switch, Route } from 'react-router-dom'
...
-    <Route path={rootPath} component={Routes} />
-    <Route exact path="/signup/createuser/:tokenstring" component={CreateUserPage} />
-    <Route component={NotFound} />
-  </Switch>
-
-export default withRouter(RootContainer)
+export const rootPath = '/2'
 
-export const Routes = ({ match }) =>
+export const RootContainer = () =>
```
Comment (matt-sm) [inline app/routes.js]: 🎉

File: app/routes.js [hunk 10]
```diff
@@ -3,24 +3,19 @@ import { withRouter, Switch, Route } from 'react-router-dom'
...
   <Switch>
-    <Route exact path={match.url} component={SignupContainer} />
+    <Route exact path={rootPath} component={SignupContainer} />
     <Route path={`${rootPath}/signup`} component={SignupContainer} />
-    <Route path={`${rootPath}/createuser/:tokenstring`} component={CreateUserPage} />
+    <Route path={`${rootPath}/createuser`} component={CreateUserPage} />
     <Route path={`${rootPath}/brief/:brief_id/respond`} component={BriefResponse} />
+    <Route path={`${rootPath}/reset-password`} component={ResetPasswordPage} />
```
Comment (matt-sm) [inline app/routes.js]: still inconsistent with our routes though

File: app/pages/ResetPasswordPage.js (hunk 7 above)
Comment (Blackfry) [reply app/pages/ResetPasswordPage.js]: the function computes form values while the prop is a straight dispatch action

File: app/actions/resetPasswordActions.js [hunk 11]
```diff
@@ -0,0 +1,94 @@
+import {
```
Comment (matt-sm) [inline app/actions/resetPasswordActions.js]: I suggest we start the fat arrow migration with these files

File: app/actions/resetPasswordActions.js [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+  return dispatch => {
+    dispatch(sendingRequest(true))
+    dmapi({ url: `/reset-password/${token}` }).then(response => {
+      if (response.error) {
+        if (response.data.message) {
+          dispatch(setErrorMessage(`${response.data.message}. Try resending reset password email.`))
+        } else {
+          dispatch(setErrorMessage(GENERAL_ERROR))
+        }
+        dispatch(handleGetResetFailure())
```
Comment (matt-sm) [inline app/actions/resetPasswordActions.js]: do you need this?

File: app/routes.js (hunk 10 above)
Comment (Blackfry) [reply app/routes.js]: hyphen?

File: app/actions/resetPasswordActions.js (hunk 12 above)
Comment (Blackfry) [reply app/actions/resetPasswordActions.js]: It's used in the resetpasswordform to toggle pagealerts

File: app/pages/CreateUserPage.js [hunk 13]
```diff
@@ -19,7 +19,10 @@ export class CreateUserPage extends BaseForm {
   }
 
   componentWillMount() {
-    const tokenString = this.props.match.params.tokenstring
+    const tokenString = this.props.location.pathname.substring(
```
Comment (matt-sm) [inline app/pages/CreateUserPage.js]: not sure - the router match on /createuser
should you call `loadRegistrationData()` if there is no token?
and would /createuser return a 404?

**Pull request merged**